File: 2023/rust/src/day1.rs
use std::fs;

// This function takes in a list of lines, all of which containing at least one
// digit. It is supposed to return the sum of the double digit numbers (calibration values)
// created by sticking together the first and last digits on each line.

pub fn part1(input_file: &str) {
    get_calibration_value(input_file, get_digit);
}

pub fn part2(input_file: &str) {
    get_calibration_value(input_file, get_digit_or_spelled_out);
}

fn get_calibration_value(input_file: &str, digit_selector_function: fn(&str) -> Option<i32>) {
    let contents = fs::read_to_string(input_file).expect("Should have been able to read the file");
    let lines = contents.split("\n");

    println!("Supplied calibration document:\n{}", contents);

    let calibration_value: i32 = lines
        .into_iter()
        .map(|line| line.trim())
        .filter(|line| line != &"")
        .map(|line| {
            let length = line.chars().count();
            let first_digit = (0..length)
                .find_map(|i| digit_selector_function(&line[i..]))
                .unwrap();
            let last_digit = (0..length)
                .rev()
                .find_map(|i| digit_selector_function(&line[i..]))
                .unwrap();
            10 * first_digit + last_digit
        })
        .sum();

    println!("Calibration value: {}", calibration_value);
}

fn get_digit_or_spelled_out(line: &str) -> Option<i32> {
    get_digit(line).or_else(|| get_spelled_out_digit(line))
}

fn get_digit(line: &str) -> Option<i32> {
    let c = line.chars().next()?;
    if c.is_digit(10) {
        Some(c.to_string().parse().unwrap())
    } else {
        None
    }
}

fn get_spelled_out_digit(line: &str) -> Option<i32> {
    let spelled_out_digits = vec![
        (0, "zero"),
        (1, "one"),
        (2, "two"),
        (3, "three"),
        (4, "four"),
        (5, "five"),
        (6, "six"),
        (7, "seven"),
        (8, "eight"),
        (9, "nine"),
    ];

    for spelled_out_digit in &spelled_out_digits {
        if line.starts_with(spelled_out_digit.1) {
            return Some(spelled_out_digit.0);
        }
    }
    return None;
}

File: python/src/2023/day1.py
from solution import Solution
from typing import List, Tuple


def filter_input(input: list):
    filtered_input = []
    for line in input:
        filtered_input.append("".join(filter(lambda x: x.isdigit(), line)))

    return filtered_input


def get_calibration_values(filtered_input: list):
    calibration_values = []
    for line in filtered_input:
        if len(line) >= 2:
            calibration_values.append(line[0] + line[len(line) - 1])
        else:
            calibration_values.append(line[0] + line[0])

    return calibration_values


class Day1(Solution):
    def first_part(self, input_file: str):
        file = open(input_file, "r")
        input = str(file.read()).splitlines()
        file.close()
        filtered_input = filter_input(input)
        calibration_values = get_calibration_values(filtered_input)
        print(sum(list(map(int, calibration_values))))

    def second_part(self, input_file: str):
        file = open(input_file, "r")
        input = str(file.read()).splitlines()
        file.close()
        calibration_values = get_calibration_values_part2(input)
        calibration_values = list(map(int, calibration_values))
        print(sum(calibration_values))


def parse_input_lists(input: str) -> Tuple[List[int], List[int]]:
    left = []
    right = []
    for line in input.splitlines():
        first, second = line.split("   ")
        left.append(int(first))
        right.append(int(second))

    return (left, right)


numbers = {
    1: "one",
    2: "two",
    3: "three",
    4: "four",
    5: "five",
    6: "six",
    7: "seven",
    8: "eight",
    9: "nine",
}


def first(line: str):
    first_digit = ""

    for j in range(len(line)):
        for i in range(1, 10):
            if line[0].isdigit():
                first_digit = line[0]
                return first_digit
            elif line.startswith(numbers[i], j) or line.startswith(str(i), j):
                first_digit = str(i)
                return first_digit


def last(line: str):
    last_digit = ""

    for j in range(len(line)):
        for i in range(1, 10):
            if line[0].isdigit():
                last_digit = line[0]
                return last_digit
            elif line.startswith("".join(reversed(numbers[i])), j) or line.startswith(
                str(i), j
            ):
                last_digit = str(i)
                return last_digit


def get_calibration_values_part2(inpt: list):
    calibration_values = []
    first_digit = ""
    last_digit = ""

    for line in inpt:
        first_digit = first(line)
        last_digit = last("".join(reversed(line)))
        calibration_values.append(first_digit + last_digit)

    return calibration_values
